Applications/Spire/Source/Ui/CalendarDatePicker.cpp

```diff
@@ -0,0 +1,114 @@
+#include "Spire/Ui/CalendarDatePicker.hpp"
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/LocalScalarValueModel.hpp"
+#include "Spire/Ui/Button.hpp"
+#include "Spire/Ui/TextBox.hpp"
+
+using namespace boost::gregorian;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+class CalendarDatePicker::MonthSpinner : public QWidget {
+  public:
+
+    explicit MonthSpinner(QWidget* parent = nullptr)
+      : MonthSpinner(
+          std::make_shared<LocalDateModel>(day_clock::local_day()), parent) {}
+
+    explicit MonthSpinner(date date, QWidget* parent = nullptr)
+      : MonthSpinner(std::make_shared<LocalDateModel>(date), parent) {}
+
+    explicit MonthSpinner(
+        std::shared_ptr<DateModel> model, QWidget* parent = nullptr)
+        : QWidget(parent),
+        m_model(std::move(model)) {
+      auto layout = new QHBoxLayout(this);
+      layout->setContentsMargins({scale_width(4), 0, scale_width(4), 0});
+      layout->setSpacing(scale_width(8));
+      const auto BUTTON_SIZE = scale(16, 16);
+      m_previous_button = make_icon_button(
+        imageFromSvg(":Icons/calendar-arrow-left.svg", BUTTON_SIZE));
+      m_previous_button->setFixedSize(BUTTON_SIZE);
+      m_previous_button->connect_clicked_signal([=] { decrement(); });
+      layout->addWidget(m_previous_button);
+      m_label = make_label("", this);
```

Comment:
Use a model.

Reply:
Do you mean something like this:

m_text_model = std::make_shared<TextModel>(get_label_text());
auto label = make_label(m_text_model, this);

Reply:
Something along the lines of a ```ToTextModel<T>```, takes a model representing a value of type ```T``` and transforms ```T``` into a ```QString```. Perhaps by default it uses the ```displayText``` to perform the conversion, with a constructor that allows a custom conversion function to be used.

Reply:
You would also possibly look at taking a ```QString``` and converting it into a ```T``` as well, since models have a ```set_current```. If an invalid ```QString``` is provided, return ```QValidator::Invalid```.

Reply:
A possible implementation: https://gist.github.com/spireken/84bd542a941e01cedb8222bec39af35e

Like the ToString function type uses displayText(...) by default, I could write a template function that takes a QString and returns a constructed object of the given template type that a default ToType method could use, but, that would mean defining how types are constructed from strings that don't have a constructor or method to construct them from a string.